### ISSUE TITLE
Brushes - Improve thumbnails

### ISSUE THREAD

**Issue by Draise14** [Bforartists/Bforartists#2233]
# Feature Request
![image](https://user-images.githubusercontent.com/35357409/107564984-16277480-6bb1-11eb-9ade-3653daf027a7.png)
Add actual relevent custom thumbnail to the brushes so it's easier to use the dropdown without reading. 

Similar to this addon:
[https://gumroad.com/l/zlbpz](https://gumroad.com/l/zlbpz)

# Task
1. Use already GPL thumbs, generate or create thumbnails
2. Setup thumbnails in the startup blend to custom

OR

Setup default thumbnails to be more relevant on compile, so default isn't the one and same thumbnail for everything. 

**Relevent Editors:**
Weight Paint Mode
Texture Paint Mode
Vertex Paint Mode

**Comment by ReinerBforartists** [comment]
We could include the addon. But i doubt that the original developer would be happy about it :D

Not sure if i am able to change this. But let's make it a task. Sounds like a great improvement. And maybe Iyad can have a look at it.

**Comment by Draise14** [comment]
We don't have to include the addon, it's a brush manager, but the thumbnails is more of a bonus. The method he implemented is just.... making the addon switch the thumbnails of the relevant brushes to "custom", then linking it to the thumbnail included with the addon. You'd then have to save the startup blend. 

![image](https://user-images.githubusercontent.com/35357409/107854549-8a009180-6dea-11eb-9020-1b7c71787b02.png)

We would just.. eh.. create custom thumbnails, set brush thumbnail to custom, save the startup. 

Unless.. these thumbnails are stored somewhere easy to modify before building? I would have to dig around.

**Comment by ReinerBforartists** [comment]
Please dig ^^

And beware of copyright of these icons. But when it's under gpl then we shouldn't have to worry.

**Comment by Draise14** [comment]
Yeah. I'm thinking of making new ones.

**Comment by ReinerBforartists** [comment]
@Draise14 , shall i assign you here?

**Comment by Draise14** [comment]
Sure, I can do that, give me a few weeks though, I am still in crunch - can't wait to get out of it and rethink my life. 

I think I also want to make another task to show/hide the thumbnail path, so the brushes thumbnail system look more streamlined.

**Comment by ReinerBforartists** [comment]
Related Brushes Panel - Toggle custom path visibility #2264

**Comment by Draise14** [comment]
Ah right.. so maybe it's not as straight forward to pack these as custom thumbs but compiled ones... I'll postpone for now.

**Comment by ReinerBforartists** [comment]
Seems so ^^

**Comment by ReinerBforartists** [comment]
The custom icons are definitely not stored in the blend file. They remain external. So you should have a look at the gumroad addon how they deal with the custom icons.

But at least the bug is fixed. The custom icon paths are not longer corrupted by saving the startup.blend

**Comment by Draise14** [comment]
Yeah. I think they store in with the same %LOCAL%/config/(build)/scripts/addons/(addon) folder that the addon extracts to when you install. 

I will do tests with custom thumbs to compile next week when I try out the new dev build. 

I assume it would have to be relative from the startup.blend and in the %LOCAL% folder somewhere. 

Do you know how Bforartists write to the local config? Where is that copied from on first use from the repository?

**Comment by ReinerBforartists** [comment]
You could have a look at the materials library fx addon. They do some path voodoo in the init.py. I needed to add our own libraries there too.

**Comment by Draise14** [comment]
Cool, I see.. when you activate the addon it technically copies files from one location to another based on the %USER% location. But it is authored from the addon location.

**Comment by Draise14** [comment]
The issue I've found here with the addon, is that I have to register it to make it read/write thumbnails. Meaning on a new installation, the thumbnails wouldn't be applied. I am taking a look at compiling in new thumbnails now.

**Comment by Draise14** [comment]
![image](https://user-images.githubusercontent.com/35357409/114246009-2fe9fb80-9957-11eb-915b-1eacbca030ae.png)
Ok.. Found it. You define the icon in `interface_icons.c` then set them to the brush in `rna_brush.c`. You then make them as png with the relevant name in `bforartists-repository\Bforartists\release\datafiles\brushicons`

Cool, then compile. Will do a compile test next week with some changes, then try add a new icon to a brush that has a repeat icon.

**Comment by Draise14** [comment]
I would have to make an addon to set custom icons to the brushes, I can't find a method to set icons from a compile. The addon would first need to register and copy it's files to %USER% before I could make the startup.blend save the files. Then the user would need to save the startup to keep the changes. 

A factory icon system is not available.

**Comment by ReinerBforartists** [comment]
I knew it's a can of worms ;)

**Comment by Draise14** [comment]
Gonna turn this into an addon to expose the brushes to the top level into the toolshelf, since I now have the code to switch brush presets. So.. changed the title to reflect that.
